tf_encrypted/keras/layers/flatten.py

```diff
@@ -0,0 +1,52 @@
+"""Flatten Layer implementation."""
+
+import numpy as np
+from tensorflow.python.keras.utils import conv_utils
+
+from tf_encrypted.keras.engine import Layer
+
+
+
+class Flatten(Layer):
+  """Flattens the input. Does not affect the batch size.
+  If inputs are shaped `(batch,)` without a channel dimension, then flattening
+  adds an extra channel dimension and output shapes are `(batch, 1)`.
+  Arguments:
+      data_format: A string,
+          one of `channels_last` (default) or `channels_first`.
+          The ordering of the dimensions in the inputs.
+          `channels_last` corresponds to inputs with shape
+          `(batch, ..., channels)` while `channels_first` corresponds to
+          inputs with shape `(batch, channels, ...)`.
+          If you never set it, then it will be "channels_last".
+  """
+
+  def __init__(self, data_format=None, **kwargs):
+    super(Flatten, self).__init__(**kwargs)
+    self.data_format = conv_utils.normalize_data_format(data_format)
+
+  def build(self, input_shape):
+    self.built = True
+
+  def call(self, inputs):
+    input_shape = inputs.shape.as_list()
+
+    if len(input_shape) == 1:
+      flatten_shape = [input_shape[0], 1]
+    else:
+      flatten_shape = [input_shape[0], -1]
+
+    outputs = self.prot.reshape(inputs, flatten_shape)
+
+    return outputs
+
+  def compute_output_shape(self, input_shape):
```

Comment:
the code in this function is broken -- the code in the if block is unreachable if input_shape is always well-defined (i.e. non-empty and not None), but the code below the if block will fail with an IndexError if input_shape is allowed to be either empty and a TypeError if input_shape is allowed to be None.

suggest you revisit the TensorShape class in TF and see if it's useful.  if so, we should discuss with @mortendahl if we should be using it more often throughout the codebase, and if not, you should figure out what the intended behavior is and simulate it more precisely with plain Python

Reply:
Thanks good catch. I am now raising an error when input_shape is empty or None. 

@mortendahl, @jvmancuso and I had a discussion about TensorShape. it comes with several functionalities when dealing with shapes. I will open an issue so we can discuss if we want to use it across the code base. Thanks

Reply:
Sounds good 👍 one problem with partial shapes is the triple stuff, since they can't be generated offline if the shape is not complete. We could make that a special case though, as the online triple source should work with partial shapes (although we may have to add some control dependencies).

Reply:
Sounds good, that totally makes sense.